pkg/controller/v1beta1/inferenceservice/components/predictor.go

```diff
@@ -241,6 +247,129 @@ func (p *Predictor) Reconcile(isvc *v1beta1.InferenceService) (ctrl.Result, erro
+		if isvc.Spec.Predictor.WorkerSpec == nil {
+			isvc.Spec.Predictor.WorkerSpec = &v1beta1.WorkerSpec{
+				PodSpec: v1beta1.PodSpec{},
+			}
+		}
+		mergedWorkerPodSpec, err := isvcutils.MergePodSpec(&sRuntime.WorkerSpec.ServingRuntimePodSpec, &isvc.Spec.Predictor.WorkerSpec.PodSpec)
+		if err != nil {
+			isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+				Reason:  v1beta1.InvalidPredictorSpec,
+				Message: "Failed to consolidate serving runtime WorkerSpec PodSpecs",
+			})
+			return ctrl.Result{}, errors.Wrapf(err, "failed to consolidate serving runtime WorkerSpec PodSpecs")
+		}
+
+		// Set the worker size from InferenceService to ServingRuntime worker size
+		if isvc.Spec.Predictor.WorkerSpec.Size != 0 {
+			sRuntime.WorkerSpec.Size = isvc.Spec.Predictor.WorkerSpec.Size
+		}
+
+		// Replace placeholders in runtime container by values from inferenceservice metadata
+		if err = isvcutils.ReplacePlaceholders(workerContainer, isvc.ObjectMeta); err != nil {
+			isvc.Status.UpdateModelTransitionStatus(v1beta1.InvalidSpec, &v1beta1.FailureInfo{
+				Reason:  v1beta1.InvalidPredictorSpec,
+				Message: "Failed to replace placeholders of worker container in serving runtime ",
+			})
+			return ctrl.Result{}, errors.Wrapf(err, "failed to replace placeholders of worker container in serving runtime")
+		}
+
+		workerPodSpec = *mergedWorkerPodSpec
+		workerPodSpec.Containers = []v1.Container{
+			*workerContainer,
+		}
+		// Label filter will be handled in ksvc_reconciler
+		sRuntimeWorkerLabels = sRuntime.WorkerSpec.ServingRuntimePodSpec.Labels
+		sRuntimeWorkerAnnotations = utils.Filter(sRuntime.WorkerSpec.ServingRuntimePodSpec.Annotations, func(key string) bool {
+			return !utils.Includes(constants.ServiceAnnotationDisallowedList, key)
+		})
+
+		//
+		workerSize := constants.DefaultWorkerMinSize
+		if sRuntime.WorkerSpec.Size > constants.DefaultWorkerMinSize {
+			workerSize = sRuntime.WorkerSpec.Size
+		}
+		// Check if the PIPELINE_PARALLEL_SIZE environment variable specified
+		// Priority for worker node count:
+		// 	1. PIPELINE_PARALLEL_SIZE environment variable
+		// 	2. WorkerSpec.Size
+		for _, container := range podSpec.Containers {
+			if container.Name == constants.InferenceServiceContainerName {
+				if envPipelineParallelSize, exists := utils.GetEnvVarValue(container.Env, constants.PipelineParallelSizeEnvName); exists {
+					// pipelineParallelSize should be bigger than 1 (head + worker)
+					if intPipelineParallelSize, err := strconv.Atoi(envPipelineParallelSize); err == nil && intPipelineParallelSize > 1 {
+						// pipelineParallelSize value is higher priority than workerSpec.Size, so pipelineParallelSize - 1(head node) will be set to workerSize
+						if intPipelineParallelSize-1 != workerSize {
+							workerSize = intPipelineParallelSize - 1
+						}
+					} else {
+						p.Log.Info("Using the WorkerSpec.Size because the provided value for pipelineParallelSize is less than 1")
```

Comment:
Move this logic to validation webhook instead and reject if user specifies `pipelineParallelSize` less than 1

Reply:
Do you think we need this logic to both servingruntime and also inferenceservice ?

Reply:
@yuzisun updated.